src/compiler/emitter.ts

```diff
@@ -961,6 +961,7 @@ export function emitFiles(
     }
 
     function markLinkedReferences(file: SourceFile) {
+        if (ts.isSourceFileJS(file)) return; // JS files don't use reference calculations as they don't do import ellision, no need to calculate it

```

Comment:
Hopefully we remember to remove this if/when decorator metadata becomes a thing again with non-experimental decorators, since preserving imports used for those types I think will end up mattering even in JS files (cross this issue with that other bug I fixed).

Actually, does this work for JSX files? or does this trigger that _other_ bug where we don't realize that `import "react"` is going to happen?

Reply:
e.g. should we copy the test from https://github.com/microsoft/TypeScript/pull/59193 and make it a .jsx file?

Reply:
In JSX files the `import "react"` we add to the output is injected by the emitter based on compiler options and doesn't depend on the refrenced-ness of the synthetic import thing in the checker (you can check pretty easily by seeing there's no `resolver.isReferencedAliasDeclaration` type calls in the jsx transform - only in the ts transform, and only if the file isn't JS).

Reply:
But yeah, should circumstances change and `shouldEmitAliasDeclaration` in `ts.ts` changes, this'd need to change to match.

Reply:
Fun fact: I tried skipping this walk for files with nothing in their `.imports` member, too (so no references to external modules), but tests showed that we *still* needed import elision calculation done even in files with no external imports, because local elision on `export import a = b` statements is done using the same logic and relies on this running to work.  And there's no simple way to know if those are present without traversing the tree anyway. 🤷‍♂️ I could have done `if (!length(file.imports) && !(file.transformFlags & TransformFlags.TypeScript)) return;` (since `export import =` is TS syntax), but how many `.ts` files don't contain any typescript syntax - hardly seems worth it.

Reply:
Ah, I guess it doesn't matter because we don't care about exact errors in this case, yeah.